cryptoservice/certificate.go

```diff
@@ -15,28 +12,11 @@ import (
 // GenerateCertificate generates an X509 Certificate from a template, given a GUN
 func GenerateCertificate(rootKey data.PrivateKey, gun string) (*x509.Certificate, error) {
 
-	algorithm := rootKey.Algorithm()
-	var (
-		publicKey  crypto.PublicKey
-		privateKey crypto.PrivateKey
-		err        error
-	)
-	switch algorithm {
-	case data.RSAKey:
-		var rsaPrivateKey *rsa.PrivateKey
-		rsaPrivateKey, err = x509.ParsePKCS1PrivateKey(rootKey.Private())
-		privateKey = rsaPrivateKey
-		publicKey = rsaPrivateKey.Public()
-	case data.ECDSAKey:
-		var ecdsaPrivateKey *ecdsa.PrivateKey
-		ecdsaPrivateKey, err = x509.ParseECPrivateKey(rootKey.Private())
-		privateKey = ecdsaPrivateKey
-		publicKey = ecdsaPrivateKey.Public()
+	switch rootKey.(type) {
+	case *data.RSAPrivateKey, *data.ECDSAPrivateKey:
+		// go doesn't fall through
 	default:
-		return nil, fmt.Errorf("only RSA or ECDSA keys are currently supported. Found: %s", algorithm)
-	}
-	if err != nil {
-		return nil, fmt.Errorf("failed to parse root key: %s (%v)", gun, err)
+		return nil, fmt.Errorf("only RSA or ECDSA keys are currently supported. Found: %s", rootKey.Algorithm())
```

Comment:
Non-blocking:  Maybe specify that they have to be bare keys, not x509 keys, since we're specifically excluding the x509 ones?